modules/sdk/project-templates/project-templates/src/main/java/com/liferay/project/templates/internal/util/ProjectTemplatesUtil.java

```diff
@@ -30,43 +37,139 @@
  */
 public class ProjectTemplatesUtil {
 
-	public static File getArchetypeFile(String artifactId, File file)
-		throws IOException {
+	public static File getArchetypeFile(String artifactId) throws IOException {
+		if (_archetypeFiles.containsKey(artifactId)) {
+			return _archetypeFiles.get(artifactId);
+		}
 
-		try (JarFile jarFile = new JarFile(file)) {
-			Enumeration<JarEntry> enumeration = jarFile.entries();
+		Properties projectTemplateJarVersions = getProjectTemplateJarVersions();
 
-			while (enumeration.hasMoreElements()) {
-				JarEntry jarEntry = enumeration.nextElement();
+		if (projectTemplateJarVersions.containsKey(artifactId)) {
+			String version = String.valueOf(
+				projectTemplateJarVersions.get(artifactId));
 
-				if (jarEntry.isDirectory()) {
-					continue;
-				}
+			try {
+				String jarName = getArchetypeJarName(artifactId, version);
 
-				String name = jarEntry.getName();
-
-				if (!name.startsWith(artifactId + "-")) {
-					continue;
-				}
+				InputStream inputStream =
+					ProjectTemplatesUtil.class.getResourceAsStream(jarName);
 
 				Path archetypePath = Files.createTempFile(
 					"temp-archetype", null);
 
 				Files.copy(
-					jarFile.getInputStream(jarEntry), archetypePath,
+					inputStream, archetypePath,
 					StandardCopyOption.REPLACE_EXISTING);
 
 				File archetypeFile = archetypePath.toFile();
 
+				_archetypeFiles.put(artifactId, archetypeFile);
+
 				archetypeFile.deleteOnExit();
 
 				return archetypeFile;
 			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
 		}
 
 		return null;
 	}
 
+	public static File getArchetypeFile(String artifactId, File file)
+		throws IOException {
+
+		File returnValue = getArchetypeFile(artifactId);
+
+		if (returnValue == null) {
+			try (JarFile jarFile = new JarFile(file)) {
+				Enumeration<JarEntry> enumeration = jarFile.entries();
+
+				while (enumeration.hasMoreElements()) {
+					JarEntry jarEntry = enumeration.nextElement();
+
+					if (jarEntry.isDirectory()) {
+						continue;
+					}
+
+					String name = jarEntry.getName();
+
+					if (!name.startsWith(artifactId + "-")) {
+						continue;
+					}
+
+					Path archetypePath = Files.createTempFile(
+						"temp-archetype", null);
+
+					Files.copy(
+						jarFile.getInputStream(jarEntry), archetypePath,
+						StandardCopyOption.REPLACE_EXISTING);
+
+					File archetypeFile = archetypePath.toFile();
+
+					_archetypeFiles.put(artifactId, archetypeFile);
+
+					archetypeFile.deleteOnExit();
+
+					returnValue = archetypeFile;
+				}
+			}
+		}
+
+		return returnValue;
+	}
+
+	public static String getArchetypeJarName(
+		String artifactId, String version) {
+
+		String jarName = "/" + artifactId + "-" + version + ".jar";
+
+		return jarName;
+	}
+
+	public static Collection<String> getArchetypeJarNames() throws IOException {
```

Comment:
For consistency with existing code in this project please change the return value from `Collection` to `List`